vyper/venom/parser.py

```diff
@@ -0,0 +1,165 @@
+from vyper.venom.context import IRContext
+from vyper.venom.basicblock import (
+    IRLabel,
+    IRVariable,
+    IRLiteral,
+    IROperand,
+    IRInstruction,
+    IRBasicBlock,
+)
+from vyper.venom.function import IRFunction
+from lark import Lark, Transformer
+
+
+VENOM_PARSER = Lark(
+    """
+    %import common.CNAME
+    %import common.DIGIT
+    %import common.LETTER
+    %import common.WS
+    %import common.INT
+
+    start: function* data_section
+    function: "function" NAME "=>" "{" block* "}"
+    data_section: "data:" call*
+
+    block: NAME ":" (call | assignment)*
+
+    assignment: VAR_IDENT "=" expr
+    expr: call | CONST
+    call: OPCODE operands_list
+
+    operands_list: (operand ("," operand)*)?
+
+    operand: VAR_IDENT | CONST | LABEL
+
+    CONST: INT
+    OPCODE: CNAME
+    VAR_IDENT: "%" INT (":" INT)?
+    LABEL: "@" NAME
+    NAME: (DIGIT|LETTER|"_")+
+
+    %ignore WS
+    """
+)
+
+
+def _set_last_var(fn: IRFunction):
+    for block in fn.get_basic_blocks():
+        output_vars = (
+            instruction.output
+            for instruction in block.instructions
+            if instruction.output is not None
+        )
+        for output_var in output_vars:
+            assert isinstance(output_var, IRVariable)
+            value = output_var.value
+            assert value.startswith("%")
+            fn.last_variable = max(fn.last_variable, int(value[1:]))
+
+
+def _set_last_label(ctx: IRContext):
+    for fn in ctx.functions.values():
+        for bb in fn.get_basic_blocks():
+            label = bb.label.value
+            if label.isdigit():
+                ctx.last_label = max(int(label), ctx.last_label)
+
+
+class VenomTransformer(Transformer):
+    def start(self, children) -> IRContext:
+        ctx = IRContext()
+        funcs = children[:-1]
+        data_section = children[-1]
+        for fn_name, blocks in funcs:
+            fn = ctx.create_function(fn_name)
+            for block_name, instructions in blocks:
+                # Get default function block if entry bb
+                if block_name == fn_name:
```

Comment:
this is a bit of a weird restriction. why does the entry basic block need to be named the same thing as the function?

Reply:
because this is the only basic block the `IRFunction` class creates by default

Reply:
i don't think we really need this, since the user might name the entry block something different than the function name. i think we should just clear the basic block dict before entering into this loop.